### ISSUE TITLE
eeprom not working with blackpill STM32f411

### ISSUE THREAD

**Issue by muhammadelmogy** [grblHAL/Plugin_EEPROM#1]
I am trying to connect eeprom to the blackpill but it is not working.
The way I am testing this by changing some settings values and disconnect  and reconnect the board again and read the value but it always revert to its default value.

I have connected 24lc32a and 24lc256 on a breadboard with the black bill, I have set A0,A1,A2 all to 0 and I have checked the address using i2c scanner with Arduino and confirmed the address was 0x50 also I have changed the address in both
eeprom_24LC16B.c
eeprom_24AAxxx.c

I have set EEPROM_ENABLE to 1 and also to 2 but nothing happen.
I have enabled I2c by adding those lines in my board.
#define I2C_ENABLE 1
#define I2C_PORT 1 // PB8 / PB9

I have changed input  pin assignment for user-control controls (cycle start, reset, feed hold) input pins since it has some conflicts with I2C port.

How exactly should I know that the firmware is aware of the eeprom?

**Comment by terjeio** [comment]
When enabling a 256kbit EEPROM these defines are needed in the map file:
```
#undef I2C_ENABLE
#define I2C_ENABLE 1
#undef EEPROM_ENABLE
#define EEPROM_ENABLE 2
#define I2C_PORT 1 // PB8 / PB9
```

Change `EEPROM_ENABLE` to 3 for 32kbit.
Alternatively enable EEPROM support in [my_machine.h](https://github.com/grblHAL/STM32F4xx/blob/2f068396d8f1cdd31cba6c7c238670a300686524/Inc/my_machine.h#L74).

Change `#define CONTROL_INMODE` to `#define CONTROL_INMODE GPIO_BITBAND`

>  I have checked the address using i2c scanner with Arduino and confirmed the address was 0x50 also I have changed the address

Do not change this as the address is shifted one bit to the right in code: 0xA0 => 0x50

> How exactly should I know that the firmware is aware of the eeprom?

`$I` will report `[NVS STORAGE:*EEPROM]` and settings should survive a cold start.

**Comment by muhammadelmogy** [comment]
Did not work!
I have cloned the code, made a new branch and made a new entry in platformio.ini file
like this

```
[env:mlabs_controller]
board = genericSTM32F411CE
board_build.ldscript = STM32F411CEUX_FLASH.ld
build_flags = ${common.build_flags}
  -D BOARD_MLABS_CONTROLLER=
  -D USB_SERIAL_CDC=1 
  -D EEPROM_ENABLE=2
lib_deps = ${common.lib_deps}
  eeprom
lib_extra_dirs = ${common.lib_extra_dirs}
; change MCU frequency
upload_protocol = dfu 
```
Further more I made a new map file by making a copy of blackpill_map.h and added required entries in several places and the code is compiling fine.
I have enabled the I2C and eeprom as described and changed some pin assignments in order to make room for the I2C port 

Currently this is the $I 

```
$I
[VER:1.1f.20220703:]
[OPT:VNMSL,35,1024,3,0]
[NEWOPT:ENUMS,RT+,TC,SED]
[FIRMWARE:grblHAL]
[NVS STORAGE:*EEPROM]
[DRIVER:STM32F411]
[DRIVER VERSION:220703]
[BOARD:Mlabs Controller]
[AUX IO:1,2,0,0]
ok
```

As you can see, [NVS STORAGE:*EEPROM] is showing that there is an eeprom but still when I change any setting and disconnect the USB cable and reconnect the setting value return back as before!

**Comment by muhammadelmogy** [comment]
I have disconnected the data lines for the eeprom(SCL and SDA) and [NVS STORAGE:*EEPROM] is still the same!
Also there was no error message to indicate the eeprom was disconnected !

**Comment by muhammadelmogy** [comment]
This is how I connect my eerpom on the breadboard
![eeprom1](https://user-images.githubusercontent.com/8148383/178152945-e4e133e0-e3ef-4dc9-a725-d17137a6779a.jpg)
![eeprom2](https://user-images.githubusercontent.com/8148383/178152948-0652562e-3dd1-45c6-8334-c15197fcbb25.jpg)

**Comment by terjeio** [comment]
I have now tested with a 256kbit EEPROM and a F411 Blackpill (plus a F303 Blackpill) and it works for me. 
Be aware that the Blackpill map uses PB9 as an AUX input, I missed that initially - and not changing that causes the I2C communication to fail. I have also tested with the `I2C1_ALT_PINMAP` (PB6/PB7) and that works too.

> Also there was no error message to indicate the eeprom was disconnected !

Error 7 is reported on startup, but that message is lost when communication is over USB. I want to delay writing to the USB port until a client is connected - but have not managed yet to make that work.

**Comment by muhammadelmogy** [comment]
I did exactly what you have mentioned and still not working!
I hooked up an oscilloscope to the SCL & SDA lines and when I connect the USB (or make a hard reset) I can see some data but after that nothing happens!! even I have tried to do for example $100=300 but nothing came on the oscilloscope, shouldn't changing some settings store the change on the eeprom immediately ?
Or the eeprom Init failed from the beginning and all further calls were just not successful after that?
How do I check this?
Where could be the failing points that I should look after and make sure there are no issue or what so ever?

**Comment by terjeio** [comment]
> changing some settings store the change on the eeprom immediately ?

Yes, but after a brief delay (microseconds) since grblHAL per default buffers writes. There are a couple of alarm conditions that blocks processing of $-commands, what does your real-time report (from a `?` command) look like? Do you get the `ok` response when changing a setting?

You may also try `$100` after setting the value to `$100=300`, this should report back the current setting.

> Where could be the failing points that I should look after and make sure there are no issue or what so ever?

A stuck bus is the easiest to check - SDA and SCL should be 3.3V when idle. The EEPROM symbol in the drawing above is a bit odd in that the WP (WRite Protect) signal is drawn as inverted (active low). All the EEPROMS I have worked with have the WP signal active high. If the symbol is correct then the WP signal should be connected to VCC instead?

There are many - bad chip, bad soldering, compiler not producing correct code due to incorrect options, ...

Zip and attach your map file and `my_machine.h` and I will generate a binary for you to test.

**Comment by muhammadelmogy** [comment]
Guess what, it is now working!!!!!
I absolutely have no idea why!!!
I really hate it when something like this happens!
One thing I made, I am using SWD for flashing because boot0 is not working for unknown reason giving me "USB device not recognized" when pressing the boot0 button, but I managed to make run once and flashed the code using DFU after that it worked just fine.
I have no idea if this is related or not!
